### PR TITLE
Add frompyfunc

### LINKED ISSUE #2956: Better Dask Array support of `frompyfunc`
NumPy's [`frompyfunc`]( https://docs.scipy.org/doc/numpy-1.13.0/reference/generated/numpy.frompyfunc.html ) is quite useful for wrapping up a basic Python function as a `ufunc`. It also works pretty much out of the box with Dask Arrays thanks to [`__array_ufunc__`]( https://docs.scipy.org/doc/numpy-1.13.0/user/basics.subclassing.html#array-ufunc-for-ufuncs ).

However there are two caveats. First one must have NumPy 1.13 at least for `__array_ufunc__` to work (we support 1.10+ currently). Second it appears that `frompyfunc` objects cannot be pickled. ( https://github.com/numpy/numpy/issues/8099 ) The latter hurts any distributed computing use cases where the `frompyfunc` object must be shipped between processes.

One way to fix both of these would be to create some sort of wrapper object derived from [Dask Array's `ufunc`]( https://github.com/dask/dask/blob/9e9fa10ef11bfe86a52214d8a2cda19508a4ee37/dask/array/ufunc.py#L45 ) that delays creating of the `frompyfunc` object until the computation is performed. This should allow the `frompyfunc` object to interact more reliably with Dask Arrays on older NumPy versions. Further it would allow the object to pickled more easily (assuming the Python function used can be pickled).

### PR DESCRIPTION
**Pull request by jcrist**
Add support for frompyfunc to dask.array.

Fixes #2956.

- [x] Tests added / passed
- [x] Passes `flake8 dask`
- [x] Fully documented, including `docs/source/changelog.rst` for all changes
      and one of the `docs/source/*-api.rst` files for new API

### CODE REVIEW

File: dask/array/tests/test_ufunc.py [hunk 1]
```diff
@@ -263,6 +267,56 @@ def test_angle():
...
+    f = da_frompyfunc(add, 2, 1)
+    np_f = np.frompyfunc(add, 2, 1)
+    x = np.array([1, 2, 3])
+
+    # Callable
+    np.testing.assert_equal(f(x, 1), np_f(x, 1))
+
+    # picklable
+    f2 = pickle.loads(pickle.dumps(f))
+    np.testing.assert_equal(f2(x, 1), np_f(x, 1))
```
Comment (jakirkham) [inline dask/array/tests/test_ufunc.py]: Thanks for testing this. :+1:

**Comment by jakirkham** [comment]
Thanks for doing this @jcrist. Looks pretty good to me.

One question that popped up when playing with `frompyfunc` is how to better determine the output type. Raised to NumPy as issue ( https://github.com/numpy/numpy/issues/10171 ). Would be curious to hear your thoughts on that issue ( https://github.com/numpy/numpy/issues/10171 ) and on this question in the Dask context as well.

**Comment by jcrist** [comment]
Since we rely on numpy to do the type inference here for the wrapped function, we can mirror whatever numpy decides upon (if anything). I see no issues in implementing the behavior here using dask.

**Comment by jakirkham** [comment]
> Since we rely on numpy to do the type inference here for the wrapped function, we can mirror whatever numpy decides upon (if anything).

Agreed.

**Pull request merged**